File: web/src/lib/eventWikipedia.ts
import { api, clusterPath } from "./api";
import {
  EVENT_CATALOG_HREF,
  eventCatalogConsumerHref,
  eventCatalogStreamHref,
  formatEventCatalogSchema,
  parseEventCatalogSchema,
  type EventCatalogConsumer,
} from "./eventCatalog";

export type EventWikipediaHistory = {
  createdAt?: string;
  updatedAt?: string;
  updatedBy?: string;
  streams: string[];
};

export type EventWikipediaDeprecation = {
  deprecated: boolean;
  successorSubject?: string;
  note?: string;
};

export type EventWikipediaIncidentLink = {
  stream: string;
  consumer: string;
  service?: string;
};

export type EventWikipediaPage = {
  subject: string;
  purpose?: string;
  history: EventWikipediaHistory;
  owner?: string;
  consumers: EventCatalogConsumer[];
  example?: Record<string, unknown> | null;
  schema?: Record<string, unknown> | null;
  relatedEvents: string[];
  knownIncidents: EventWikipediaIncidentLink[];
  deprecation: EventWikipediaDeprecation;
  documented: boolean;
  orphan: boolean;
};

export type EventWikipediaTotals = {
  total: number;
  documented: number;
  deprecated: number;
  orphan: number;
};

export type EventWikipediaSnapshot = {
  capturedAt?: string;
  pages: EventWikipediaPage[];
  totals: EventWikipediaTotals;
};

const EMPTY_TOTALS: EventWikipediaTotals = {
  total: 0,
  documented: 0,
  deprecated: 0,
  orphan: 0,
};

export const EVENT_WIKIPEDIA_HREF = "/docs/event-wikipedia";

export async function fetchEventWikipedia(
  clusterId: string,
  options?: { fresh?: boolean; subject?: string },
): Promise<EventWikipediaSnapshot> {
  const params = new URLSearchParams();
  if (options?.fresh) params.set("fresh", "1");
  if (options?.subject?.trim()) params.set("subject", options.subject.trim());
  const qs = params.toString() ? `?${params.toString()}` : "";
  const snap = await api<EventWikipediaSnapshot>(clusterPath(clusterId, `/event-wikipedia${qs}`));
  const data = snap.data;
  return {
    capturedAt: data?.capturedAt,
    pages: Array.isArray(data?.pages) ? data.pages : [],
    totals: data?.totals ?? EMPTY_TOTALS,
  };
}

export function filterEventWikipediaPages(
  pages: EventWikipediaPage[],
  query: string,
): EventWikipediaPage[] {
  const q = query.trim().toLowerCase();
  if (!q) return pages;
  return pages.filter((p) => {
    const hay = [
      p.subject,
      p.owner ?? "",
      p.purpose ?? "",
      p.deprecation.successorSubject ?? "",
      ...(p.history.streams ?? []),
      ...(p.relatedEvents ?? []),
    ]
      .join(" ")
      .toLowerCase();
    return hay.includes(q);
  });
}

export function sortEventWikipediaPages(pages: EventWikipediaPage[]): EventWikipediaPage[] {
  return [...pages].sort((a, b) => a.subject.localeCompare(b.subject));
}

export function eventWikipediaCatalogHref(subject: string): string {
  return `${EVENT_CATALOG_HREF}?q=${encodeURIComponent(subject)}`;
}

export function eventWikipediaSubjectHref(subject: string): string {
  return `${EVENT_WIKIPEDIA_HREF}?subject=${encodeURIComponent(subject)}`;
}

export function eventWikipediaIncidentHref(
  clusterId: string,
  incident: EventWikipediaIncidentLink,
): string {
  const params = new URLSearchParams({
    cluster: clusterId,
    stream: incident.stream,
    consumer: incident.consumer,
  });
  return `/admin/audit?${params.toString()}`;
}

export {
  EVENT_CATALOG_HREF,
  eventCatalogConsumerHref,
  eventCatalogStreamHref,
  formatEventCatalogSchema,
  parseEventCatalogSchema,
};

File: internal/domain/event_wikipedia.go
package domain

import (
	"sort"
	"strings"
	"time"

	commonstrings "github.com/gopherust-io/nats-consol/pkg/common/strings"
)

// EventWikipediaHistory is auto-assembled history for a subject page.
type EventWikipediaHistory struct {
	CreatedAt time.Time `json:"createdAt,omitzero"`
	UpdatedAt time.Time `json:"updatedAt,omitzero"`
	UpdatedBy string    `json:"updatedBy,omitempty"`
	Streams   []string  `json:"streams"`
}

// EventWikipediaDeprecation is curated deprecation status.
// goalign:ignore // JSON DTO; trailing bool padding is unavoidable
type EventWikipediaDeprecation struct {
	SuccessorSubject string `json:"successorSubject,omitempty"`
	Note             string `json:"note,omitempty"`
	Deprecated       bool   `json:"deprecated"`
}

// EventWikipediaIncidentLink is a deep-link target for consumer incident reconstruction.
type EventWikipediaIncidentLink struct {
	Stream   string `json:"stream"`
	Consumer string `json:"consumer"`
	Service  string `json:"service,omitempty"`
}

// EventWikipediaPage is one auto-assembled Wikipedia article for a subject.
type EventWikipediaPage struct {
	Deprecation    EventWikipediaDeprecation    `json:"deprecation"`
	Subject        string                       `json:"subject"`
	Purpose        string                       `json:"purpose,omitempty"`
	Owner          string                       `json:"owner,omitempty"`
	History        EventWikipediaHistory        `json:"history"`
	Consumers      []EventCatalogConsumer       `json:"consumers"`
	Example        []byte                       `json:"example,omitempty"`
	Schema         []byte                       `json:"schema,omitempty"`
	RelatedEvents  []string                     `json:"relatedEvents"`
	KnownIncidents []EventWikipediaIncidentLink `json:"knownIncidents"`
	Documented     bool                         `json:"documented"`
	Orphan         bool                         `json:"orphan"`
}

// EventWikipediaTotals summarizes Wikipedia inventory.
type EventWikipediaTotals struct {
	Total      int `json:"total"`
	Documented int `json:"documented"`
	Deprecated int `json:"deprecated"`
	Orphan     int `json:"orphan"`
}

// EventWikipediaSnapshot is the assembled Docs Wikipedia response.
type EventWikipediaSnapshot struct {
	CapturedAt time.Time            `json:"capturedAt"`
	Pages      []EventWikipediaPage `json:"pages"`
	Totals     EventWikipediaTotals `json:"totals"`
}

// BuildEventWikipedia merges catalog entries with genome peers into Wikipedia pages.
// Read-only assembly for Docs UI: no JetStream writes; consumers/streams are
// displayed from live inventory, and deprecation/examples stay in Postgres.
func BuildEventWikipedia(catalog EventCatalogSnapshot, genome EventGenomeSnapshot) EventWikipediaSnapshot {
	relatedBySubject := genomeRelatedBySubject(genome)

	pages := make([]EventWikipediaPage, 0, len(catalog.Entries))
	var totals EventWikipediaTotals
	for _, e := range catalog.Entries {
		related := relatedBySubject[e.Subject]
		if related == nil {
			related = []string{}
		}
		incidents := make([]EventWikipediaIncidentLink, 0, len(e.Consumers))
		for _, c := range e.Consumers {
			incidents = append(incidents, EventWikipediaIncidentLink{
				Stream:   c.Stream,
				Consumer: c.Name,
				Service:  c.Service,
			})
		}
		consumers := e.Consumers
		if consumers == nil {
			consumers = []EventCatalogConsumer{}
		}
		streams := e.Streams
		if streams == nil {
			streams = []string{}
		}
		page := EventWikipediaPage{
			Subject: e.Subject,
			Purpose: e.Description,
			History: EventWikipediaHistory{
				CreatedAt: e.CreatedAt,
				UpdatedAt: e.UpdatedAt,
				UpdatedBy: e.UpdatedBy,
				Streams:   streams,
			},
			Owner:          e.Owner,
			Consumers:      consumers,
			Example:        e.Example,
			Schema:         e.Schema,
			RelatedEvents:  related,
			KnownIncidents: incidents,
			Deprecation: EventWikipediaDeprecation{
				Deprecated:       e.Deprecated,
				SuccessorSubject: e.SuccessorSubject,
				Note:             e.DeprecationNote,
			},
			Documented: e.Documented,
			Orphan:     e.Orphan,
		}
		if page.Documented {
			totals.Documented++
		}
		if page.Deprecation.Deprecated {
			totals.Deprecated++
		}
		if page.Orphan {
			totals.Orphan++
		}
		pages = append(pages, page)
	}
	sort.Slice(pages, func(i, j int) bool {
		return pages[i].Subject < pages[j].Subject
	})
	totals.Total = len(pages)
	return EventWikipediaSnapshot{
		CapturedAt: catalog.CapturedAt,
		Pages:      pages,
		Totals:     totals,
	}
}

// FilterEventWikipediaPages keeps pages whose subject equals subject (trimmed).
func FilterEventWikipediaPages(snap EventWikipediaSnapshot, subject string) EventWikipediaSnapshot {
	subject = strings.TrimSpace(subject)
	if commonstrings.IsEmpty(subject) {
		return snap
	}
	out := make([]EventWikipediaPage, 0, 1)
	for _, p := range snap.Pages {
		if p.Subject == subject {
			out = append(out, p)
			break
		}
	}
	snap.Pages = out
	var totals EventWikipediaTotals
	for _, p := range out {
		totals.Total++
		if p.Documented {
			totals.Documented++
		}
		if p.Deprecation.Deprecated {
			totals.Deprecated++
		}
		if p.Orphan {
			totals.Orphan++
		}
	}
	snap.Totals = totals
	return snap
}

func genomeRelatedBySubject(genome EventGenomeSnapshot) map[string][]string {
	clusters := make(map[string][]string)
	for _, f := range genome.Findings {
		key := strings.TrimSpace(f.Genome)
		if commonstrings.IsEmpty(key) {
			key = strings.TrimSpace(f.Subject)
		}
		if commonstrings.IsEmpty(key) {
			continue
		}
		if _, ok := clusters[key]; ok {
			continue
		}
		literals := make([]string, 0, len(f.Cluster))
		seen := make(map[string]struct{})
		for _, peer := range f.Cluster {
			peer = strings.TrimSpace(peer)
			if commonstrings.IsEmpty(peer) {
				continue
			}
			if _, ok := seen[peer]; ok {
				continue
			}
			seen[peer] = struct{}{}
			literals = append(literals, peer)
		}
		sort.Strings(literals)
		if len(literals) >= 2 {
			clusters[key] = literals
		}
	}

	out := make(map[string][]string)
	for _, literals := range clusters {
		for _, subject := range literals {
			peers := make([]string, 0, len(literals)-1)
			for _, peer := range literals {
				if peer != subject {
					peers = append(peers, peer)
				}
			}
			out[subject] = peers
		}
	}
	return out
}
